new blog post 2020-11-02-quarkus-native-on-aarch64 (#753)
quarkusio/quarkusio.github.io: _posts/2020-11-02-quarkus-native-on-a-raspberry-pi.adoc

### Building tools

For this PoC I've used a QEmu-base VM. That promotes reusability, but it's still expensive in terms of usability (QEmu works like a charm on linux OS, but not on Windows and MacOS). As stated at the beginning of this article, a more flexible, portable and scalable solution still based on cpu emulation and container technology is under investiogation at the moment. That would help the delegation and scalability of build process. 

Be aware that QEmu-based tools are still a workaround. GraalVM is missing the cross-compile feature and it will take a while to release it.

### Quarkus everywhere

It's clear that Quarkus can potentially run everywhere. It is, of course, a matter of supportability from GraalVM to the unterlying architecture.

So far I feel comfortable in stating that this is just the beginning and lots more features and capabilities are yet to come.  

> It would very nice to have a distribution of Mandrel for aarch64. Mandrel is aligned with the `native-image` capabilities from GraalVM with OpenJDK and Red Hat Enterprise Linux lbiraries to improve maintainability. Looking forward to testing it ;-)

Reviewer: ```suggestion
It would very nice to have a distribution of Mandrel for aarch64. Mandrel is aligned with the `native-image` capabilities from GraalVM with OpenJDK and Red Hat Enterprise Linux libraries to improve maintainability. Looking forward to testing it ;-)
```

As I’m running Fedora Workstation on my desktop, I’ve picked up the 3rd option.

### Container registry
The remote container registry of choice is https://quay.io/[Quay.io]. More details about setting up the account created on quay.io will be shared in a dedicated section of this article.

## Implementation

This section will guide you through the following 

. Setting up the VM
. Setting up the RPi and Fedora IoT
. Connecting the container engine to the quay.io account
. Compiling and deploying your Quarkus native app on Podman
. Outcome

## 1. Setting up the VM

This goal can be achieved using VM technology based on QEmu.
QEmu plays an important role in this task, cause that’s the best upstream tool to use to make a virtual machine emulating the target cpu architecture.

### Setup a Virtual machine using QEmu

On a linux distro, the use of VMM (Virtual Machine Manager is highly recommended):

[source,sh]
----
$ sudo dnf groupinstall virtualization
$ sudo dnf install qemu-system-aarch64
----

The minimum requirements for the virtual machine are the following:

* CPU: 4 cores *Cortex-A53* (ARMv8) 64-bit
* RAM: 8 GB
* Disk: 10Gb
* Networking: bridge

> * OS: Fedora Minimal (latest stable) - aarch64 (https://download.fedoraproject.org/pub/fedora-secondary/releases/32/Spins/aarch64/images/Fedora-Minimal-32-1.6.aarch64.raw.xz[link])

Reviewer: Maybe update to Fedora 33 or is it too soon?

image::/assets/images/posts/quarkus-native-on-a-raspberry-pi/native-quarkus-getting-started-on-arm8-web-page.png[width=100%]

## Conclusions

### Big Picture

This article reproduces the steps to compile and run a Quarkus native app on a Raspberry Pi cpu. This process is potentially valid for all the cpus belonging to the ARM v8 family, with some potential exceptions due to the fact the architecture of few ARM based cpus may differ from the standard one for that version.

### Building tools

For this PoC I've used a QEmu-base VM. That promotes reusability, but it's still expensive in terms of usability (QEmu works like a charm on linux OS, but not on Windows and MacOS). As stated at the beginning of this article, a more flexible, portable and scalable solution still based on cpu emulation and container technology is under investiogation at the moment. That would help the delegation and scalability of build process. 

Be aware that QEmu-based tools are still a workaround. GraalVM is missing the cross-compile feature and it will take a while to release it.

### Quarkus everywhere

> It's clear that Quarkus can potentially run everywhere. It is, of course, a matter of supportability from GraalVM to the unterlying architecture.

Reviewer 2: ```suggestion
It's clear that Quarkus can potentially run everywhere. It is, of course, a matter of supportability from GraalVM to the underlying architecture.
```

So far I feel comfortable in stating that this is just the beginning and lots more features and capabilities are yet to come.  

> It would very nice to have a distribution of Mandrel for aarch64. Mandrel (the Red Hat distribution of GraalVM) is optimized for Quarkus and makes it even more fast and reliable. Looking forward to testing it ;-)

Reviewer 2: ```suggestion
It would very nice to have a distribution of Mandrel for aarch64. Mandrel is aligned with the `native-image` capabilities from GraalVM with OpenJDK and Red Hat Enterprise Linux lbiraries to improve maintainability. Looking forward to testing it ;-)
```

mandrel is not a red hat distribution, its a GraalVM project (graal/mandrel) and yes it exists for reliability/maintainabiliy but it shouldn't (at this stage) be faster nor slower than normal GraalVM. Changed text above is taken from graal/mandrel about text as a suggestion.
